### ISSUE TITLE
Add template / formatted data binding

### ISSUE THREAD

**Issue by IntelOrca** [Basssiiie/OpenRCT2-FlexUI#29]
**Idea**
Allow labels to have a template string.

**Additional context**
Label text is often constructed using a formatted string. Currently the whole text needs to be bind to a string store, then the update method or whatever sets the store to the result of a `context.formatString`. This could be done as part of the data binding.

```ts
const model = {
    numGuests: flex.store(0)
};

...

flex.label({
    text: {
        template: 'Number of guests: {BLACK}{COMMA16}',
        args: [model.numGuests]
    }
});

... or ...

flex.label({
    text: flex.template('Number of guests: {BLACK}{COMMA16}', model.numGuests)
});
```

**Current work around**:
```ts
const model = {
    numGuests: flex.store(0)
};
const viewModel = {
    numGuests: new FormattedStore('Number of guests: {BLACK}{COMMA16}', model.numGuests)
};

...

flex.label({
    text: viewModel.numGuests
});

...

class FormattedStore<T> implements flex.Store<string> {
    private readonly fmt: string;
    readonly store: flex.Store<T>;

    constructor(fmt: string, store: flex.Store<T>) {
        this.fmt = fmt;
        this.store = store;
    }

    get() {
        return context.formatString(this.fmt, this.store.get());
    }

    subscribe(callback: (value: string) => void): () => void {
        return this.store.subscribe(value => {
            callback(this.get());
        });
    }
}
```

**Comment by IntelOrca** [comment]
This is actually a lot simpler that I expected. flex UI could simply provide a helper function that creates this facade.

```ts
class FormattedStore implements flex.Store<string> {
    private readonly fmt: string;
    readonly stores: flex.Store<any>[];

    constructor(fmt: string, ...stores: flex.Store<any>[]) {
        this.fmt = fmt;
        this.stores = stores;
    }

    get() {
        return context.formatString(this.fmt, ...this.stores.map(x => x.get()));
    }

    subscribe(callback: (value: string) => void): () => void {
        const subscriptions = this.stores
            .map(x => x.subscribe(_ => callback(this.get())));
        return () => subscriptions.every(x => x());
    }
}

function bindTemplate(fmt: string, ...args: flex.Store<T>[]) {
    return new FormattedStore(fmt, ...args);
}
```

```ts
flex.label({
    text: bindTemplate('Number of guests: {BLACK}{COMMA16}', model.numGuests)
}),
```

**Comment by Basssiiie** [comment]
Thank you for your feature requests. 😃 

Regarding this specific one, I haven't really used `formatString` as I've mostly been formatting using the ES5 JS functions, but I like the idea. Wouldn't this suffice as well though?

```ts
flex.label({
    text: compute(model.numGuests, guests => context.formatString('Number of guests: {BLACK}{COMMA16}', guests))
}),
```

It supports multiple stores as well.

**Comment by IntelOrca** [comment]
The `bindTemplate` example is a nice shorthand, maybe that could use `compute` for the implementation?

> I've mostly been formatting using the ES5 JS functions

If you want to format lengths, currency, temperature, dates etc. according to the player's locale, you will want to use `formatString`.

**Comment by Basssiiie** [comment]
Fixed as of c342c84. :)
